docs/proposals/TLSPolicy.md

```diff
@@ -0,0 +1,84 @@
+# WIP
+# TLSPolicy API
+Currently, TLS certificates issued by the MGC are issued and managed using `cert-manager`. Configuration for these certificates are opaque to the end-user - we'd like to gradually expose a TLSPolicy to cover a range of use cases where administrators require more flexible configuration.
+
+## Problem
+
+Gateway Administrators require a comprehensive way to manage Automated, ACME-based TLS for Gateways. As gateways multiply and scale, there's a need for a flexible, policy-based API for handling Automatic TLS. Currently (July 2023), the MGC sets up automated TLS whenever a Gateway listener is created. There are currently no hooks into the creation of automated TLS certificates, nor user-exposed configuration options in terms ACME providers, custom renewal windows/schedules, key size, algorithm and other configuration options available via `cert-manager`.
+
+## Use Cases
+
+**Choosing a TLS Provider**: As an admin, I want to select a specific ACME provider for the TLS certificates in a particular gateway.
+
+**Setting the TLS Certificate Renewal Window**: As an admin, I need to define the renewal period for TLS certificates (e.g., every 90 days) and a lead time for renewals (e.g., 15 days before expiration).
+
+**Choosing a TLS Strategy**: As an admin, I want to employ different strategies for managing certificates (e.g., multiple host certificate or per host certificate).
+
+**Choosing a TLS Strategy**: As an admin, I want to be able to apply enterprise security standards to certificates generated by the MGC. I want to be able to specify certificate key size, algorithm and encoding.
+
+
+## Current implementation
+
+
+
+## Goals
+
+- Design an API that allows admins to handle Automatic TLS for any given gateway.
+- Facilitate easy selection of a TLS provider.
+- Allow admins to set the renewal window for TLS certificates.
+- Enable the use of different TLS strategies.
+- Consider additional TLS configurations such as enforcing HTTPS listeners and supporting wildcard hosts.
```

Comment:
I think it should be a goal that TLS is not managed unless a TLSPolicy is specified. By default TLS will be the responsibility of the gateway admin. If they want us to take that on they can create a TLSPolicy.
Perhaps not part of this, but we prob need to handle listeners that are not backed by a MZ IE something directly handled by the admin for a CNAME for example or self managed DNS. 
@mikenairn interested in your opinion here? We might not have to start here but we will need to handle a listener that is not backed by a zone in a graceful way

Reply:
I would need to go and check what we currently do for TLS, but I'm not seeing where ManagedZones comes into it. DNSPOlicy does require that records are created in a known hosted zone, but the TLSPolicy wouldn't since it would be (i expect) created with it's own cert manager Issuer reference, with provider credentials added in the usual cert manager issuer way to deal with challenges. 

Maybe I'm missing the point, quite possible.

Reply:
Yeah good point, so when setting up cert-manager issuer you would provide it with a credential. The listeners hosts would have to be in the same DNS provider as with the current design only one issuer per gateway would be allowed. But I agree that we don't need to tie to the existence of a managed zone

Reply:
I see, yes we are limiting it to only one listener. I think this comes back to the issue that we have already talked about and being able to apply polices to listeners, rather than just at the Gateway level.

I guess we could hook it into the ManagedZone somehow :thinking:, although I'm not totally sure how that would work, or if we really want that.

Reply:
There is potentially a complex relationship here
1) I choose an issuer in the TLSPolicy and that applies to all listeners
2) some of the listeners are backed by a zone in a different provider 

I think having to have a managed zone setup before using a policy may be the wrong user experience https://github.com/orgs/Kuadrant/discussions/7